### DISCUSSION TITLE
oath app doesn't work?

### DISCUSSION THREAD

**Discussion by lgxz** [solokeys/solo2-cli#37] [General]
`
% solo2 app oath list

thread 'main' panicked at 'explicit panic', src/transport/ctap.rs:90:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

`

**Comment by provi1** [comment]
Hi,
i used solo v2 normal version with .8 firmware on windows 10 with solo2-v0.1.1-x86_64 cli 
i tested oath - register, list, delete and totp all worked fine.
but i have only tested the list parameter with one registered seceret.

**Reply by lgxz** [reply]
Tested with .9 firmware, still panicked at transport/ctap.rs:

`
impl From<u8> for Code {
    fn from(code: u8) -> Self {
        use Code::*;
        match code {
            0x1 => Ping,
            0x6 => Init,
            0x8 => Wink,
            0x3F => Error,
            0x3B => Keepalive,
            vendor_code @ 0x40..=0x7F => Vendor(VendorCode::new(vendor_code)),
            _ => panic!(),
        }
    }
}
`

My OS is MacOS 12.2

Edit:
Tested on Ubuntu 20.04.03, same error.

**Reply by nickray** [reply]
Yeah, the OATH app doesn't work over CTAP, you need PCSC. It'd need translation of the various methods to CTAPHID vendor extensions, I'm not planning to do this.

I think unfortunately Ubuntu 20.04 (and macOS) doesn't have a recent enough CCID driver. For Ubuntu you might sideload a newer version (or overwrite Info.plist), for macOS other users have managed to update the driver: https://github.com/solokeys/solo2-cli/discussions/19. This is a problem that should eventually solve itself as updates trickle in. A more radical solution is to switch distro to something that stays up to date like Arch or Fedora :innocent: 

At the core of this is that Linux CCID only speaks to allowlisted USB VID:PID pairs (whereas on Windows, for once, everything works out of the box). So we relatively early sent a PR to add ourselves, and then later there was a release, and so on...

**Comment by loveshack** [comment]
For anyone hitting this on Debian-based systems, and possibly other GNU/Linux, I put a solution in https://github.com/solokeys/solo2-cli/issues/52#issuecomment-1054548831
